mesonbuild/envconfig.py

```diff
@@ -83,31 +83,69 @@
-
-                section[entry] = res
-
-            out[s] = section
-        return out
+    def __init__(self, filenames: T.List[str]):
+        from .coredata import CmdLineFileParser
+        self.parser = CmdLineFileParser()
+        self.constants = {'True': True, 'False': False}
+        self.sections = {}
+
+        self.parser.read(filenames)
+
+        # Parse [constants] first so they can be used in other sections
+        if self.parser.has_section('constants'):
+            self.constants.update(self.parse_section('constants'))
+
+        for s in self.parser.sections():
+            if s == 'constants':
+                continue
+            self.sections[s] = self.parse_section(s)
+
+    def parse_section(self, s):
+        self.scope = self.constants.copy()
+        section = {}
+        for entry in self.parser[s]:
+            if ' ' in entry or '\t' in entry or "'" in entry or '"' in entry:
+                raise EnvironmentException('Malformed variable name {!r} in machine file.'.format(entry))
+            value = self.parser[s][entry]
+            # Windows paths...
+            value = value.replace('\\', '\\\\')
+            try:
+                ast = mparser.Parser(value, 'machinefile').parse()
+                res = self.evaluate_statement(ast.lines[0])
+            except MesonException:
+                raise EnvironmentException('Malformed value in machine file variable {!r}.'.format(entry))
+            except KeyError as e:
+                raise EnvironmentException('Undefined constant {!r} in machine file variable {!r}.'.format(e.args[0], entry))
+            section[entry] = res
+            self.scope[entry] = res
+        return section
+
+    def evaluate_statement(self, node):
+        if isinstance(node, (mparser.StringNode)):
+            return node.value
+        elif isinstance(node, mparser.BooleanNode):
+            return node.value
+        elif isinstance(node, mparser.NumberNode):
+            return node.value
+        elif isinstance(node, mparser.ArrayNode):
+            return [self.evaluate_statement(arg) for arg in node.args.arguments]
+        elif isinstance(node, mparser.IdNode):
+            return self.scope[node.value]
+        elif isinstance(node, mparser.ArithmeticNode):
+            l = self.evaluate_statement(node.left)
+            r = self.evaluate_statement(node.right)
+            if node.operation == 'add':
+                if (isinstance(l, str) and isinstance(r, str)) or \
+                   (isinstance(l, list) and isinstance(r, list)):
+                    return l + r
+            elif node.operation == 'div':
+                if isinstance(l, str) and isinstance(r, str):
+                    return os.path.join(l, r)
+        raise EnvironmentException('Unsupported node type')
+
+    @staticmethod
+    def from_config_parser(filenames: T.List[str]) -> T.Dict[str, T.Dict[str, T.Dict[str, str]]]:
+        return MesonConfigFile(filenames).sections
```

Comment:
This should be a class method and return `cls(filesnames).sections`

Reply:
From a design point of view though this is kinda icky, class that has two methods (that should be private/protected), some member variables, and is only called via a static method that instantiates an instance of itself, and returns a member. I think it would be nicer to make this a function. Even more importantly, this really belongs in coredata not in envconfig, as that avoids the need for the local import.

Reply:
Right, restructured all that now.